File: grafo.py
# grafo.py - representación del grafo y lectura de los ficheros .gr y .co

class Grafo:
    def __init__(self, nombre_mapa):
        """
        nombre_mapa: cadena base del mapa, por ejemplo 'USA-road-d.BAY'
        Leerá 'USA-road-d.BAY.co' y 'USA-road-d.BAY.gr'
        """
        self.nombre_mapa = nombre_mapa

        # Diccionario de adyacencia: id -> lista de (vecino, coste)
        self.adj = {}

        # Coordenadas: id -> (lon, lat)
        self.coords = {}

        # Contadores útiles
        self.num_vertices = 0
        self.num_arcos = 0

        # Cargamos los dos ficheros
        self._leer_co(nombre_mapa + ".co")
        self._leer_gr(nombre_mapa + ".gr")

    def _leer_co(self, filename):
        """
        Lee el fichero .co y guarda las coordenadas de cada vértice.
        Formato típico:
        v <id> <longitud> <latitud>
        """
        with open(filename, "r") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue  # línea vacía
                if not line.startswith("v"):
                    # ignoramos comentarios u otras líneas
                    continue

                partes = line.split()
                # Ejemplo: ["v", "1", "-122.3456", "37.1234"]
                _, vid, lon, lat = partes

                vid = int(vid)
                lon = float(lon)
                lat = float(lat)

                self.coords[vid] = (lon, lat)

        self.num_vertices = len(self.coords)

    def _leer_gr(self, filename):
        """
        Lee el fichero .gr y guarda los arcos con su coste.
        Formato típico:
        a <origen> <destino> <coste>
        """
        with open(filename, "r") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                if not line.startswith("a"):
                    # ignoramos comentarios, cabeceras, etc.
                    continue

                partes = line.split()
                # Ejemplo: ["a", "1", "2", "1988"]
                _, u, v, coste = partes

                u = int(u)
                v = int(v)
                coste = int(coste)

                # Si el nodo no está en adj, lo creamos
                if u not in self.adj:
                    self.adj[u] = []
                if v not in self.adj:
                    self.adj[v] = []  # así sabemos que existe el vértice

                # Añadimos el arco dirigido u -> v
                self.adj[u].append((v, coste))
                self.num_arcos += 1

    def vecinos(self, v):
        """
        Devuelve la lista de vecinos de v en forma [(hijo, coste), ...].
        Si el vértice no tiene salida, devuelve lista vacía.
        """
        return self.adj.get(v, [])

    def tiene_coordenadas(self, v):
        """
        True si tenemos lon/lat para el vértice v.
        """
        return v in self.coords

    def coordenadas(self, v):
        """
        Devuelve (lon, lat) del vértice v, o None si no existe.
        """
        return self.coords.get(v, None)

    def __str__(self):
        """
        Representación simple del grafo (para debug rápido).
        """
        return f"Grafo({self.nombre_mapa}): {self.num_vertices} vertices, {self.num_arcos} arcos"

File: grafo.cpp
#include "grafo.hpp"
#include <fstream>
#include <sstream>
#include <iostream>
#include <cmath>
#include <algorithm>

Grafo::Grafo() : num_arcos_cargados(0), max_id_nodos(0) {}

bool Grafo::cargar(const std::string& nombre_mapa) {
    std::string fichero_gr = nombre_mapa + ".gr";
    std::string fichero_co = nombre_mapa + ".co";

    // ---------------------------------------------------------
    // 1. Lectura del fichero de GRAFO (.gr)
    // ---------------------------------------------------------
    std::ifstream file_gr(fichero_gr);
    if (!file_gr.is_open()) {
        std::cerr << "Error: No se pudo abrir el fichero de grafo: " << fichero_gr << std::endl;
        return false;
    }

    std::string linea;
    // Reservamos un tamaño inicial razonable para evitar redimensionamientos constantes al principio.
    // Los grafos DIMACS suelen ser grandes.
    lista_adyacencia.reserve(100000); 

    while (std::getline(file_gr, linea)) {
        if (linea.empty()) continue;

        // Solo nos interesan las líneas que empiezan por 'a'
        if (linea[0] == 'a') {
            std::stringstream ss(linea);
            char tipo;
            int u, v, coste;
            
            // Formato: a <id1> <id2> <coste>
            ss >> tipo >> u >> v >> coste;

            // Ajuste dinámico del tamaño del vector si encontramos un ID nuevo más alto
            int max_idx = std::max(u, v);
            if (max_idx >= lista_adyacencia.size()) {
                lista_adyacencia.resize(max_idx + 1);
            }
            if (max_idx > max_id_nodos) {
                max_id_nodos = max_idx;
            }

            // Añadir arista (Grafo Dirigido)
            lista_adyacencia[u].push_back({v, coste});
            num_arcos_cargados++;
        }
        // Ignoramos líneas 'p', 'c', etc.
    }
    file_gr.close();

    // ---------------------------------------------------------
    // 2. Lectura del fichero de COORDENADAS (.co)
    // ---------------------------------------------------------
    std::ifstream file_co(fichero_co);
    if (!file_co.is_open()) {
        std::cerr << "Error: No se pudo abrir el fichero de coordenadas: " << fichero_co << std::endl;
        return false;
    }

    // Aseguramos que el vector de coordenadas tenga tamaño suficiente
    if (max_id_nodos >= coordenadas.size()) {
        coordenadas.resize(max_id_nodos + 1);
    }

    long long nodos_coord_leidos = 0;
    while (std::getline(file_co, linea)) {
        if (linea.empty()) continue;

        // Solo nos interesan las líneas que empiezan por 'v'
        if (linea[0] == 'v') {
            std::stringstream ss(linea);
            char tipo;
            int id;
            double lon_int, lat_int; // Leemos como double para simplificar, aunque el input es entero
            
            // Formato: v <id> <longitud> <latitud>
            ss >> tipo >> id >> lon_int >> lat_int;

            // Si el ID es mayor que lo que vimos en el grafo (raro, pero posible), redimensionamos
            if (id >= coordenadas.size()) {
                coordenadas.resize(id + 1);
                // También redimensionamos la lista de adyacencia para mantener consistencia de índices
                if (id >= lista_adyacencia.size()) {
                    lista_adyacencia.resize(id + 1);
                }
                max_id_nodos = std::max(max_id_nodos, id);
            }

            // Guardar convertidos a coordenadas reales (dividir por 10^6 según enunciado)
            // NOTA: El enunciado dice longitud primero, luego latitud.
            coordenadas[id] = {lat_int / 1000000.0, lon_int / 1000000.0};
            nodos_coord_leidos++;
        }
    }
    file_co.close();

    // Mostrar resumen de carga (útil para debug y cumplir requisitos de salida por pantalla)
    std::cout << "# vertices cargados (coords): " << nodos_coord_leidos << std::endl;
    std::cout << "# arcos cargados: " << num_arcos_cargados << std::endl;

    return true;
}

const std::vector<Arista>& Grafo::get_vecinos(int id_nodo) const {
    // Verificación de seguridad básica
    if (id_nodo < 0 || id_nodo >= lista_adyacencia.size()) {
        static const std::vector<Arista> vacio;
        return vacio;
    }
    return lista_adyacencia[id_nodo];
}

Punto Grafo::get_coordenadas(int id_nodo) const {
    if (id_nodo < 0 || id_nodo >= coordenadas.size()) {
        return {0.0, 0.0}; // Retorno por defecto si no existe
    }
    return coordenadas[id_nodo];
}

int Grafo::get_num_nodos() const {
    return max_id_nodos; // O coordenadas.size(), dependiendo de la definición exacta
}

long long Grafo::get_num_arcos() const {
    return num_arcos_cargados;
}

bool Grafo::existe_nodo(int id) const {
    return id >= 0 && id < lista_adyacencia.size();
}